src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -57,6 +59,14 @@ def _build_x509_name(backend, x509_name):
     return x509.Name(attributes)
 
 
+def _build_general_name(backend, gn):
+    if gn.type == backend._lib.GEN_DNS:
+        data = backend._ffi.buffer(
+            gn.d.dNSName.data, gn.d.dNSName.length
+        )[:].decode("ascii")
+        return x509.DNSName(idna.decode(data))
```

Comment:
Ok, allow me to walk through this:

* We get a bytes object out of OpenSSL
* We ASCII decode that -> the only things stored in X.509 certs are ascii strings which have been IDNA encoded. We now have a unicode text object containing something IDNA encoded
* We decode *again*, which makes sense because idna is a text to text encoding
* In the end we have a unicode string which expresses all parts of the codepoint space that are relevant to IDNs

Confirm?

(Also, can someone who makes poor life decisions, such as @glyph, @exarkun, or @hynek also confirm :-))